// npm-upgrade/lib/cliTable.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

exports.createSimpleTable = createSimpleTable;

var _cliTable = require('cli-table');

var _cliTable2 = _interopRequireDefault(_cliTable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const COL_ALIGNS_MAP = {
  l: 'left',
  r: 'right',
  c: 'middle'
};

function createSimpleTable(rows, opts = {}) {
  if (opts.colAligns) {
    opts.colAligns = opts.colAligns.split('').map(val => COL_ALIGNS_MAP[val]);
  }

  const table = new _cliTable2.default((0, _extends3.default)({
    style: { 'padding-left': 2 },
    colAligns: ['left', 'right', 'right', 'right', 'middle'],
    chars: {
      'top': '',
      'top-mid': '',
      'top-left': '',
      'top-right': '',
      'bottom': '',
      'bottom-mid': '',
      'bottom-left': '',
      'bottom-right': '',
      'left': '',
      'left-mid': '',
      'mid': '',
      'mid-mid': '',
      'right': '',
      'right-mid': '',
      'middle': ''
    }
  }, opts));

  if (rows) {
    table.push(...rows);
  }

  return table;
}
